File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0636_Exclusive_Time_of_Functions.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

public class Question_0636_Exclusive_Time_of_Functions {
    public static class Log {
        public int id;
        public boolean isStart;
        public int time;
        public int subDuration;

        public Log(String content) {
            String[] strs = content.split(":");
            id = Integer.parseInt(strs[0]);
            isStart = strs[1].equals("start");
            time = Integer.parseInt(strs[2]);
            subDuration = 0;
        }
    }

    public int[] exclusiveTime(int n, List<String> logs) {
        int[] res = new int[n];
        Deque<Log> stack = new ArrayDeque<>();
        for (String content : logs) {
            Log log = new Log(content);
            if (log.isStart) {
                stack.push(log);
            } else {
                Log top = stack.pop();
                res[top.id] += (log.time - top.time + 1 - top.subDuration);
                if (!stack.isEmpty()) {
                    stack.peek().subDuration += (log.time - top.time + 1);
                }
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0636_Exclusive_Time_of_Functions.cc
#include <vector>
#include <stack>
#include <sstream>

using namespace std;

struct Log {
  int id;
  bool isStart;
  int time;
  int subDuration;
};

class Solution {
 public:
  vector<int> exclusiveTime(int n, vector<string>& logs) {
    vector<int> res(n, 0);
    stack<Log> st;
    for(string log: logs) {
      stringstream ss(log);
      string temp, temp2, temp3;
      getline(ss, temp, ':');
      getline(ss, temp2, ':');
      getline(ss, temp3, ':');

      Log item = {stoi(temp), temp2 == "start", stoi(temp3), 0};
      if (item.isStart) {
        st.push(item);
      } else {
        int t = item.time - st.top().time + 1;
        res[st.top().id] += (t - st.top().subDuration);
        st.pop();
        if (!st.empty()) {
          st.top().subDuration += t;
        }
      }
    }
    return res;
  }
};
